src/services/refactors/moveToFile.ts

```diff
@@ -0,0 +1,1114 @@
+        addNewFileToTsconfig(program, changes, oldFile.fileName, newFile, hostGetCanonicalFileName(host));
+    }
+    else {
+        const sourceFile = program.getSourceFile(newFile);
+        if (sourceFile) {
+            const importAdder = codefix.createImportAdder(sourceFile, context.program, context.preferences, context.host);
+            getNewStatementsAndRemoveFromOldFile(oldFile, newFile, usage, changes, toMove, program, host, newFile, preferences, /*newFileExists*/ true, importAdder);
+        }
+    }
+}
+
+function getNewStatementsAndRemoveFromOldFile(
+    oldFile: SourceFile, newFile: string, usage: UsageInfo, changes: textChanges.ChangeTracker, toMove: ToMove, program: Program, host: LanguageServiceHost, newFilename: string, preferences: UserPreferences, newFileExists: boolean, importAdder?: codefix.ImportAdder
+) {
+    const checker = program.getTypeChecker();
+    const prologueDirectives = takeWhile(oldFile.statements, isPrologueDirective);
+    if (oldFile.externalModuleIndicator === undefined && oldFile.commonJsModuleIndicator === undefined && usage.oldImportsNeededByNewFile.size === 0 && usage.newFileImportsFromOldFile.size === 0 && !newFileExists) {
+        deleteMovedStatements(oldFile, toMove.ranges, changes);
+        return [...prologueDirectives, ...toMove.all];
+    }
+
+    let useEsModuleSyntax = true;
+    if (newFileExists) {
+        const newFilePath = program.getSourceFile(newFile)?.path;
+        if (extensionFromPath(newFile) === Extension.Js && newFilePath && getImpliedNodeFormatForFile(newFilePath, /*packageJsonInfoCache*/ undefined, host, program.getCompilerOptions()) !== ModuleKind.ESNext && (oldFile.commonJsModuleIndicator || program.getCompilerOptions().verbatimModuleSyntax)) {
+            useEsModuleSyntax = false;
+        }
+    }
+    const quotePreference = getQuotePreference(oldFile, preferences);
+    const importsFromNewFile = createOldFileImportsFromNewFile(oldFile, usage.oldFileImportsFromNewFile, newFilename, program, host, useEsModuleSyntax, quotePreference);
+    if (importsFromNewFile) {
+        insertImports(changes, oldFile, importsFromNewFile, /*blankLineBetween*/ true, preferences);
+    }
+
+    deleteUnusedOldImports(oldFile, toMove.all, changes, usage.unusedImportsFromOldFile, checker);
+    deleteMovedStatements(oldFile, toMove.ranges, changes);
+    updateImportsInOtherFiles(changes, program, host, oldFile, usage.movedSymbols, newFilename, quotePreference);
+
+    const imports = getNewFileImportsAndAddExportInOldFile(oldFile, newFile, usage.oldImportsNeededByNewFile, usage.newFileImportsFromOldFile, changes, checker, program, host, useEsModuleSyntax, quotePreference, importAdder);
+    const body = addExports(oldFile, toMove.all, usage.oldFileImportsFromNewFile, useEsModuleSyntax);
+    if (newFileExists) {
+        const sourceFile = program.getSourceFile(newFile);
+        if (sourceFile && sourceFile.statements.length > 0) {

```

Comment:
Style nit:

If `sourceFile` is `undefined`, something has gone horribly wrong since you were told that `newFileExists` and you shouldn’t continue. Wrap `program.getSourceFile(newFile)` in a `Debug.checkDefined` or something. And since you access it in an earlier `if` block, just do that at the very top of the function. (`getSourceFile` is pretty cheap but not free, and it’s nice to avoid repeating yourself.)

```ts
const targetSourceFile = newFileExists ? Debug.checkDefined(program.getSourceFile(newFile)) : undefined;
```

Then, you never have to check `newFileExists` again; you can just check `if (targetSourceFile)` and know it means the same thing, while narrowing `undefined` out of its type.